### PR TITLE
Extend Cplex python API, introduce MST cplex files and logic to keep track of best bound

### PR DESCRIPTION
**Pull request by Joao-Alves**
In this pull request I introduced the following changes:

- Extend the CPLEX python API to support the nthreads parameters - This would allow developers to control the CPU usage of Cplex within Pulp
- Created support to use MST cplex solution files which provide more information when using warmStarts by keeping track of multiple solutions 
- Introduced the bestbound parameters in the LP problem - This would allow to measure the gap of the problem when a timelimit is applied and create additional logic with it

### CODE REVIEW

**Comment by Joao-Alves** [comment]
@pchtsp are you the one who should check this pull request?

**Review by pchtsp — CHANGES_REQUESTED** [review]
Thanks again for the PR! I left some questions / comments.

File: pulp/apis/cplex_api.py [hunk 1]
```diff
@@ -292,17 +292,21 @@ def __init__(
             timeLimit=None,
             gapRel=None,
             warmStart=False,
+            solutionFile=None,
             logPath=None,
+            nthreads=0,
```
[the remark this answers is not in the document]
Comment (Joao-Alves) [reply pulp/apis/cplex_api.py:297]: Ok I will change to the threads parameter

**Comment by pchtsp** [comment]
with regards to the MIP-start file. As somewhat documented in [this page](https://coin-or.github.io/pulp/guides/how_to_configure_solvers.html#using-solver-specific-functionality),  you can in theory already do this with the current access to the low-level interface to CPLEX_PY:

```python
import pulp
x = pulp.LpVariable('x', lowBound=0)
prob = pulp.LpProblem('name', pulp.LpMinimize)
prob += x
solver = pulp.CPLEX_PY()
status = prob.solve(solver)
prob.solverModel.MIP_starts.write(SOME_MST_FILE)
```

and then:

```python
import pulp
x = pulp.LpVariable('x', lowBound=0)
prob = pulp.LpProblem('name', pulp.LpMinimize)
prob += x
solver = pulp.CPLEX_PY()

solver.buildSolverModel(prob)
# you load the MST file:
solver.solverModel.MIP_starts.read(SOME_MST_FILE)
# the, you can call the solver to solve the problem
solver.callSolver(prob)
# finally, you fill the PuLP variables with the solution
status = solver.findSolutionValues(prob)

```
I haven't tested everything here, but it *should* work. If not, it should be somewhat possible tweaking some things.

-----

Alternatively, the "pulp-way" to do the same thing, should be:


```python
import pulp
x = pulp.LpVariable('x', lowBound=0)
prob = pulp.LpProblem('name', pulp.LpMinimize)
prob += x
solver = pulp.CPLEX_PY()
status = prob.solve(solver)

# here we have the variables loaded inside the solver.
# we can extract them to a dictionary of values.
VALUE_FOR_VARIABLE = {'x': x.value()}

```
and then when you want to warm start a similar problem:

```python
import pulp
x = pulp.LpVariable('x', lowBound=0)
prob = pulp.LpProblem('name', pulp.LpMinimize)
prob += x
x.setInitialValue(VALUE_FOR_VARIABLE['x'])

solver = pulp.CPLEX_PY(warmStart=True)
status = prob.solve(solver)

```

---

I personally do not see the need of adding yet another way of giving a `warmStart` to the CPLEX_PY. But I could be mistaken. What do you think?

**Comment by Joao-Alves** [comment]
I agree with you that is possible with the low level access and the example in the documentantion show how to achieve it. However, I think it requires a low level knowledge on how Pulp is implemented and interfaces with solvers, and this parameter would allow for a simpler way to achieve it. 
As well as a low level way to understand how CPLEX works.

This parameter could also be useful in Gurobi as it supports files as well even more formats then Cplex.

**Comment by pchtsp** [comment]
Hello again, I'm sorry I don't think it's a good idea to add another way to warmStart that only works for some solvers when there's ways to do that that are already documented. I understand why you do so and why it works for you though. 

If you want to do a PR with only the `threads` argument for CPLEX, I will accept it.

Thanks again for your help.

**Pull request closed without merging**